drivers/sensor/lsm6dsr/lsm6dsr.c

```diff
@@ -0,0 +1,754 @@
+/* lsm6dsr.c - Driver for LSM6DSR accelerometer, gyroscope and
+ * temperature sensor
+ */
+
+/*
+ * Copyright (c) 2023, Andrei Hutanu
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Datasheet:
+ * https://www.st.com/resource/en/datasheet/lsm6dsr.pdf
+ */
+
+#define DT_DRV_COMPAT st_lsm6dsr
+
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/init.h>
+#include <string.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/__assert.h>
+#include <zephyr/logging/log.h>
+#include "lsm6dsr.h"
+
+LOG_MODULE_REGISTER(LSM6DSR, CONFIG_SENSOR_LOG_LEVEL);
+
+static const uint16_t lsm6dsr_odr_map[] = {0, 12, 26, 52, 104, 208, 416, 833,
+					1660, 3330, 6660};
+
+static int lsm6dsr_freq_to_odr_val(uint16_t freq)
+{
+	size_t i;
+
+	for (i = 0; i < ARRAY_SIZE(lsm6dsr_odr_map); i++) {
+		if (freq == lsm6dsr_odr_map[i]) {
+			return i;
+		}
+	}
+
+	return -EINVAL;
+}
+
+static int lsm6dsr_odr_to_freq_val(uint16_t odr)
+{
+	/* for valid index, return value from map */
+	if (odr < ARRAY_SIZE(lsm6dsr_odr_map)) {
+		return lsm6dsr_odr_map[odr];
+	}
+
+	/* invalid index, return last entry */
+	return lsm6dsr_odr_map[ARRAY_SIZE(lsm6dsr_odr_map) - 1];
+}
+
+static const uint16_t lsm6dsr_gyro_fs_map[] = {245, 500, 1000, 2000, 125};
```

Comment:
wrong, it should be at index 0, 1, 2, 3, 4, 5, ... :
250, 4000, 125. 0, 500. 0, 0, 0, 1000, 0, 0, 0, 2000
i.e. range == 0 --> 250dps, range == 1 --> 4000 dps, and so on

Moreover, there are too many holes though. Maybe you can rid of this map structure.